### DISCUSSION TITLE
Facilitate precompilation for `juliacall` calls

### DISCUSSION THREAD

**Discussion by ggoretkin-bdai** [cjdoris/PythonCall.jl#473] [Q&A]
I am using `juliacall` in Python to call a Julia function. I would like to precompile the Julia function. Here are the specializations of the function by introspecting on the Python program:

```
jl.MethodAnalysis.methodinstances(jl.FooPkg.bar_func)


Julia:
2-element Vector{Core.MethodInstance}:
 MethodInstance for FooPkg.bar_func(::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::Any, ::FooPkg.InverseParams)
 MethodInstance for FooPkg.bar_func(::PyArray{Float64, 1, true, true, Float64}, ::PyArray{Float64, 1, true, true, Float64}, ::PyArray{Float64, 1, true, true, Float64}, ::PyArray{Float64, 1, true, true, Float64}, ::PyArray{Float64, 1, true, true, Float64}, ::PyArray{Float64, 1, true, true, Float64}, ::PyArray{Float64, 1, true, true, Float64}, ::PyArray{Float64, 1, true, true, Float64}, ::PyArray{Float64, 1, true, true, Float64}, ::PyArray{Float64, 1, true, true, Float64}, ::Float64, ::PyList{Any}, ::PyList{Any}, ::FooPkg.InverseParams)
 ```

I would like to basically insert the following into `FooPkg.jl`

```jl
@setup_workload begin
    ENV["JULIA_CONDAPKG_BACKEND"] = "Null" # use the system installation of Python
    using PythonCall
    np = pyimport("numpy")

    arg_1 = np.array([0.0, 0.0, 0.0])
    
    @compile_workload begin
        # all calls in this block will be precompiled, regardless of whether
        # they belong to this package or not (on Julia 1.8 and higher)
        bar_func(arg_1, ...)
    end
end
```

But this is not quite right, as `typeof(arg_1)` is `Py`, and not `PyArray{Float64, 1, true, true, Float64}`. It appears I can get a value of the right type by doing

```jl
julia> typeof(PyArray([1.0]))
PyArray{Float64, 1, true, true, Float64}
```

but is there a way to get it more directly from the corresponding Python code? More generally, what is a workload I can put in `@compile_workload` that exercises the same code paths as is exercised by `juliacall`?

**Comment by goretkin** [comment]
Also asked here: https://discourse.julialang.org/t/precompiling-for-pythoncall-jl-juliacall/111596

**Comment by cjdoris** [comment]
If you already have some Python code then using `pyexec` to execute it will be the simplest way to ensure you get the same code paths.

In your example you could do `pyjl(bar_func)(arg_1, ...)` might work. Or try `bar_func(pyconvert(Any, arg_1), ...)`.

**Reply by ggoretkin-bdai** [reply]
Thanks for the reply. I'm not sure how to proceed with using `pyexec`. I tried

```
julia> pyexec(@NamedTuple{ans}, "ans=f(p, x)", Main, (f=FlexIK.srinv, p=FlexIK.InverseParams(1.0, 1.0), x=rand(2,2)))
@NamedTuple{ans}(([0.31979155022045447 0.3069743518191882; 0.05635637500792721 0.25767403266495903],))
```

I also tried

`pyjl` on the function is causing compilation to happen for argument type `Any` and `Matrix{Float64}`, but not `PyArray{Float64, 1, true, true, Float64}`


`pyconvert` on the argument doesn't seem like it would help

```
julia> pyconvert(Any, ones(2,))
2-element Vector{Float64}:
 1.0
 1.0
```

After those evaluations, I see

```
julia> MethodAnalysis.methodinstances(FlexIK.srinv)
2-element Vector{Core.MethodInstance}:
 MethodInstance for FlexIK.srinv(::FlexIK.InverseParams, ::Matrix{Float64})
 MethodInstance for FlexIK.srinv(::FlexIK.InverseParams, ::Any)
```

Compare this with running the following Python program

```
from juliacall import Main as jl
import numpy as np


jl.seval("using FlexIK")
jl.seval("using MethodAnalysis")

print(jl.MethodAnalysis.methodinstances(jl.FlexIK.srinv))


jl.FlexIK.srinv(
    jl.FlexIK.InverseParams(1.0, 1.0),
    np.random.rand(2, 2)
)

print(jl.MethodAnalysis.methodinstances(jl.FlexIK.srinv))
```

which generates the following output

```
Core.MethodInstance[]
Core.MethodInstance[MethodInstance for FlexIK.srinv(::FlexIK.InverseParams, ::Any), MethodInstance for FlexIK.srinv(::FlexIK.InverseParams, ::PyArray{Float64, 2, true, false, Float64})]
```

**Reply by cjdoris** [reply]
What I meant with pyexec was to execute a piece of pure python code where all the data is loaded in python the way you intend to do it when using your wrapper package - and specifically NOT by passing in data from Julia, since this is the bit that's a little trickier to get right.

With the pyconvert suggestion, you tried it on `ones(2)` but in your original code `arg_1` was a numpy array - which if you evaluate `pyconvert(Any, arg_1)` you'll get a `PyArray` as I think you're intending.

**Reply by ggoretkin-bdai** [reply]
Ah, yes, that makes sense. So I can get a list of method instances hit now, but I don't see a way forward for doing this integrated into the Julia package's precompilation step.

I put up an example here: https://github.com/goretkin/PythonPrecompileExample.jl The README there shows all the method instances that should be precompiled. But uncommenting this section: https://github.com/goretkin/PythonPrecompileExample.jl/blob/7f03fd7a74cc542712d587dc1adab4561267d285/src/PythonCallPrecompileExample.jl#L84-L102 gives the error message pasted into the comments.

**Reply by cjdoris** [reply]
The error message explains it:
```
ERROR: LoadError: Python: Julia: Creating a new global in closed module `Main` (`PythonCallPrecompileExample`) breaks incremental compilation because the side effects will not be permanent.
```

You're making changes to Main and you're not allowed. It's because you did
```python
from juliacall import Main as jl
```
which is fine for interactive use or scripts but for anything else you should do
```python
import juliacall
jl = juliacall.newmodule("SomeName")
```

**Reply by ggoretkin-bdai** [reply]
Thanks for the help interpreting the error message. At first I thought it was related to `pyexec(py_script, Main)`, which is not the case.


with `jl = juliacall.Main` the error is 

```
ERROR: LoadError: Python: Julia: Creating a new global in closed module `Main` (`PythonCallPrecompileExample`) breaks incremental compilation because the side effects will not be permanent.
Stacktrace:
```

with `jl = juliacall.newmodule("NonMainModule")` the error is 

```
ERROR: LoadError: Python: Julia: Creating a new global in closed module `Main` (`#1#2`) breaks incremental compilation because the side effects will not be permanent.
```

My guess is that somewhere an anonymous function is created in `Main`, still.


I want to make sure I am not reinventing the wheel here, because this is certainly not straightforward! Is this the right technique for triggering precompilation? Is there a pre-existing example you can point me to?

**Reply by ggoretkin-bdai** [reply]
This is the anonymous function (#1#2), it appears: https://github.com/JuliaPy/PythonCall.jl/blob/6975f7c5d1f08fc4db785e40d1041492346a21c1/pysrc/juliacall/__init__.py#L12

**Reply by cjdoris** [reply]
Hmm ok. I can't totally remember why I made it a submodule of Main. You could do `jl = juliacall.Base.Module()` to get a new anonymous module?

**Reply by ggoretkin-bdai** [reply]
I think that works in the latest PythonCall.jl `v0.9.16`. On the previous version (which Pkg is choosing because in our environment, we have Julia `v1.10.2`), I get:

```
ERROR: LoadError: Python: Julia: Evaluation into the closed module `anonymous` breaks incremental compilation because the side effects will not be permanent. This is likely due to some other module mutating `anonymous` with `eval` during precompilation - don't do this.
```

Thanks for the suggestion. Ideally Julia v1.10.3 gets released soon. Do you have a reference to the specific upstream issue?

**Reply by cjdoris** [reply]
I'm not sure what upstream issue you mean - I don't think the version of Julia/JuliaCall/PythonCall should affect the behaviour you're seeing.

**Reply by ggoretkin-bdai** [reply]
> I'm not sure what upstream issue you mean

https://github.com/JuliaPy/PythonCall.jl/blob/f586f2494432f2e5366ff1e2876b8aa532630b54/docs/src/releasenotes.md#0916-2024-03-14

> Removed compatibility with Julia 1.10.1 and 1.10.2 (to be fixed in 1.10.3 and 1.11.0) due to an upstream bug

**Reply by cjdoris** [reply]
Ah #464
